Fastapi backend

chr1st1ank opened the pull request:
This PR adds a backend for Fastapi including tests and example.

In the first commits there are also fixes for 
- fixes a few smaller issues in the flask tests which only became apparent to me when I tried to add a second test suite.
- correction of a few dead links in the README.md
- some more tests for flask (request log was untested)

bobbui commented:
wow, will take a deep look soon!

bobbui commented on example/fastapi_sample_app.py:
```diff
@@ -0,0 +1,55 @@
+app = fastapi.FastAPI()
+
+# init the logger as usual
+logger = logging.getLogger(__name__)
```
you're missing the **logger.setLevel** call

chr1st1ank replied:
This is actually not necessary. All what's needed is configured further down with a config dict.
I'm doing it that way, so that I can pass this configuration to uvicorn which configures logging itself. When I tried it locally in the script as for the other backends it didn't work out.

bobbui replied:
a bit weird, when i run the sample directly from terminal the logging statement didnt emit anything, only works with setLevel call

chr1st1ank replied:
Ah, when you copy the code directly into an IPython shell, you mean? I also tested it and indeed that didn't work without `logger.setLevel()`. I only tested to run the script as dedicated python process (`python fastapi_sample_app.py`).

But right, with an additional `logger.SetLevel()` it works under all circumstances, so I just added that.

bobbui commented on json_logging/framework/fastapi/__init__.py:
```diff
@@ -0,0 +1,137 @@
+        return False
+
+
+if is_fastapi_present():
```
this method is declared twice

chr1st1ank replied:
No, the second one is an if statement using the function. This patten I adopted from other backends. 
But your comment reminds me of this mechanism and that I did not stick to the lazy importing it was for. I have to adjust the way I do the imports. Otherwise json-logging can only be used if fastapi is installed.
Will take me the one or other day until I find the time. 
So thanks for pointing me at it!

bobbui replied:
its total fine, somehow i misunderstood it LOL

chr1st1ank commented:
@bobbui, as I wrote above your comment made me aware that the fastapi implementation always imported `fastapi` and its basis `starlette`. That's probably not what we want for non-web users or for users of different frameworks.
So I changed this by putting the implementation in a separate module which is only imported if fastapi is available.

In order to guarantee that such things are caught by the tests in the future I added another test suite which runs all the examples from the README file in a minimal environment for the respective backend. E.g. the flask backend is only run in an environment with "flask" and the test tooling in it. That's the reason why I had to make it a separte job in the github action workflow. A more elegant solution did not come to my mind.
Additional benefit of these tests is that they guarantee that the README examples at least not crash immediately.
Please take a look if you like this setup.

bobbui commented:
> @bobbui, as I wrote above your comment made me aware that the fastapi implementation always imported `fastapi` and its basis `starlette`. That's probably not what we want for non-web users or for users of different frameworks.
> So I changed this by putting the implementation in a separate module which is only imported if fastapi is available.
> 
> In order to guarantee that such things are caught by the tests in the future I added another test suite which runs all the examples from the README file in a minimal environment for the respective backend. E.g. the flask backend is only run in an environment with "flask" and the test tooling in it. That's the reason why I had to make it a separte job in the github action workflow. A more elegant solution did not come to my mind.
> Additional benefit of these tests is that they guarantee that the README examples at least not crash immediately.
> Please take a look if you like this setup.

Its should be ok that way even though a bit undesirable. The integration test approach is sound, if we can somehow run each sample in a minimal isolated environment, e.g: Flask sample only run with Python virtualenv with only Flask installed. That would be really great.

chr1st1ank commented:
> 
> Its should be ok that way even though a bit undesirable. The integration test approach is sound, if we can somehow run each sample in a minimal isolated environment, e.g: Flask sample only run with Python virtualenv with only Flask installed. That would be really great.

That's just what I implemented in `.github/workflows/code_quality.yml`. It installs these minimal environments for each of these smoketests.

bobbui reviewed (approved):
LGTM

bobbui commented:
released 1.3.0
